Fix comparisions for ConcreteModuleType

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32256 Fix comparisions for ConcreteModuleType**

Poison any concrete module type created directly from a JIT type since its ConcreteModuleBuilder information is not filled in so it cannot participate in deduplication checks. This is a workaround to allow for wrapper modules of loaded things. The long-term solution is to remove the de-normalization of how we store descriptions of types.

Differential Revision: [D19423055](https://our.internmc.facebook.com/intern/diff/D19423055/)

zdevito committed:
Fix comparisions for ConcreteModuleType

Previously two unrelated modules loaded from torch.jit.load
would compare equal because we only considered their data_ attributes which
are initialized blank in torch.jit.load. This changes ConcreteModuleType
to distinguish when the data_ attribute is blank vs when it is empty.

This replaces the poisoned logic.

Differential Revision: [D19423055](https://our.internmc.facebook.com/intern/diff/D19423055/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 0cda33b1:


**None of the build failures appear to be your fault**.

* **8/8** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=0cda33b1907be33295cf9009582b41d44fb79f18).

### :snowflake: 8 failures recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4263636) pytorch_linux_xenial_py3_clang5_mobile_build (1/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86724956) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86724956)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4263643) pytorch_linux_xenial_py3_clang5_asan_build (2/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86725038) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86725038)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4263642) pytorch_linux_xenial_py2_7_9_build (3/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86725052) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86725052)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py2.7.9:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py2.7.9:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py2.7.9:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4263659) pytorch_linux_xenial_cuda9_cudnn7_py3_build (4/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86725067) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86725067)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda9-cudnn7-py3:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda9-cudnn7-py3:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda9-cudnn7-py3:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4263641) pytorch_linux_xenial_py3_clang5_android_ndk_r19c_mobile_code_analysis (5/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86725083) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86725083)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4263663) pytorch_linux_xenial_py3_clang5_android_ndk_r19c_x86_32_build (6/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86725100) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86725100)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4263652) pytorch_linux_xenial_py3_6_gcc5_4_build (7/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86725101) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86725101)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-gcc5.4:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-gcc5.4:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-gcc5.4:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4263653) pytorch_xla_linux_xenial_py3_6_clang7_build (8/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86725138) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86725138)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-clang7:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-clang7:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-clang7:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zdevito committed:
Update on "Fix comparisions for ConcreteModuleType"


Previously two unrelated modules loaded from torch.jit.load
would compare equal because we only considered their data_ attributes which
are initialized blank in torch.jit.load. This changes ConcreteModuleType
to distinguish when the data_ attribute is blank vs when it is empty.

This replaces the poisoned logic.

Differential Revision: [D19423055](https://our.internmc.facebook.com/intern/diff/D19423055/)

[ghstack-poisoned]

zdevito committed:
Update on "Fix comparisions for ConcreteModuleType"


Previously two unrelated modules loaded from torch.jit.load
would compare equal because we only considered their data_ attributes which
are initialized blank in torch.jit.load. This changes ConcreteModuleType
to distinguish when the data_ attribute is blank vs when it is empty.

This replaces the poisoned logic.

Differential Revision: [D19423055](https://our.internmc.facebook.com/intern/diff/D19423055/)

[ghstack-poisoned]

zdevito committed:
Update on "Fix comparisions for ConcreteModuleType"


Previously two unrelated modules loaded from torch.jit.load
would compare equal because we only considered their data_ attributes which
are initialized blank in torch.jit.load. This changes ConcreteModuleType
to distinguish when the data_ attribute is blank vs when it is empty.

This replaces the poisoned logic.

Differential Revision: [D19423055](https://our.internmc.facebook.com/intern/diff/D19423055/)

[ghstack-poisoned]

zdevito committed:
Update on "Fix comparisions for ConcreteModuleType"


Previously two unrelated modules loaded from torch.jit.load
would compare equal because we only considered their data_ attributes which
are initialized blank in torch.jit.load. This changes ConcreteModuleType
to distinguish when the data_ attribute is blank vs when it is empty.

This replaces the poisoned logic.

Differential Revision: [D19423055](https://our.internmc.facebook.com/intern/diff/D19423055/)

[ghstack-poisoned]
